<a href="https://colab.research.google.com/github/Mohamad-Selawy/Data-Analysis/blob/main/Data%20Analysis/Level%201/4_Data_Quality_%26_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Quality & Data Cleaning with Pandas

Welcome to this lesson on Data Quality and Data Cleaning using the powerful Pandas library in Python!

In the world of Data Analytics, raw data is rarely perfect. It often comes with imperfections, inconsistencies, and errors that can significantly impact your analysis and the conclusions you draw. Learning to identify and resolve these issues is a fundamental skill for any aspiring Data Analyst.

This notebook is designed to guide you through the process of inspecting, diagnosing, and solving common data quality problems. We will explore various techniques to clean and prepare your data, ensuring it's ready for accurate and meaningful analysis.

---

## What is Data Quality?

Data quality describes how suitable data is for analysis and decision-making. High-quality data is accurate, complete, consistent, and relevant, providing a reliable foundation for insights. Low-quality data, on the other hand, can lead to incorrect conclusions, flawed models, and poor business decisions.

Let's look at some common dimensions of data quality:

*   **Accuracy**: The data is correct and reflects the real-world facts it's supposed to represent.
*   **Completeness**: All necessary data is present, with no missing values where information should exist.
*   **Consistency**: Data values are consistent across the dataset and follow defined rules, without contradictions.
*   **Validity**: Data conforms to defined format, type, and range rules (e.g., age cannot be negative).
*   **Uniqueness**: There are no duplicate records that should represent a single entity.
*   **Timeliness**: The data is up-to-date and available when needed.

---

> Real-world datasets are rarely perfectly clean. A Data Analyst must first understand the quality of the data before performing analysis.

### The Data Cleaning Workflow

Data cleaning is not a single step but a systematic process. Here’s a typical workflow:

```text
Raw Data
   ↓
Data Inspection  🔎
   ↓
Data Quality Assessment  ⚠️
   ↓
Identify Problems
   ↓
Clean / Transform  🧹
   ↓
Validate  ✅
   ↓
Clean Dataset
   ↓
Analysis
```

---

### Why should we inspect data before starting analysis?

Inspecting data before analysis is crucial for several reasons:

1.  **Understand Dataset Structure**: It helps analysts grasp the layout, columns, and overall organization of the data.
2.  **Identify Quality Problems**: It reveals issues like missing values, duplicates, or incorrect data types early on.
3.  **Avoid Incorrect Assumptions**: You won't make assumptions about the data's cleanliness or structure that could lead to errors.
4.  **Choose Appropriate Cleaning Methods**: Understanding the nature of the problems allows you to select the most effective cleaning strategies.
5.  **Protect Analysis Accuracy**: Ensures that subsequent analysis and modeling are built on reliable data, leading to more accurate and trustworthy insights.

Let's get started by setting up our environment and loading our practice dataset!

## 2. 📚 Import Libraries

We'll start by importing the essential Python libraries that we will use throughout this notebook. Each library serves a specific purpose in our data cleaning journey.

In [1]:
import pandas as pd # For efficient data manipulation and analysis, especially with DataFrames.

## 3. 💾 Load the Dataset

We will load our practice dataset, `Data_Quality_and_Cleaning.csv`, into a Pandas DataFrame. This DataFrame will be our primary tool for exploring and cleaning data quality issues.

In [16]:
df = pd.read_csv('Data_Quality_and_Cleaning.csv')
print("Dataset loaded successfully!")

Dataset loaded successfully!


### 🔎 Initial Look: `df.head()` and `df.tail()`

The `.head()` method is used to display the first 5 rows of the DataFrame, giving you a quick overview of its structure, columns, and the type of data it contains. Conversely, `.tail()` shows the last 5 rows, which can be useful for spotting any trailing issues or summaries.


In [3]:
print("First 5 rows of the DataFrame (df.head()):")
display(df.head())

print("\nLast 5 rows of the DataFrame (df.tail()):")
display(df.tail())

First 5 rows of the DataFrame (df.head()):


,EmployeeID,Name,Department,Age,Gender,Salary,Experience,City,PerformanceScore
0,101.0,Alice Smith,IT,26.0,Female,60278.0,22.0,New York,71.0
1,NaN,Bob Johnson,HR,32.0,Male,87849.0,NaN,London,94.0
2,103.0,Charlie Brown,Finance,57.0,Male,83428.0,5.0,Paris,940.0
3,104.0,Diana Prince,MARKETING,47.0,Female,65282.0,26.0,Tokyo,86.0
4,105.0,Eve Adams,IT,42.0,Female,62734.0,290.0,Berlin,NaN



Last 5 rows of the DataFrame (df.tail()):


,EmployeeID,Name,Department,Age,Gender,Salary,Experience,City,PerformanceScore
17,118.0,Rachel Scott,HR,52.0,Female,NaN,22.0,Paris,94.0
18,119.0,Steve Turner,Marketing,27.0,MALE,104756.0,19.0,Tokyo,91.0
19,120.0,Tina Wright,Sales,40.0,Female,103158.0,27.0,Berlin,74.0
20,113.0,Mia Clark,HR,44.0,Female,57225.0,29.0,Paris,71.0
21,109.0,Ivy Davis,Marketing,NaN,NaN,106375.0,28.0,Tokyo,92.0


### 📊 Basic Dimensions: `df.shape` and `df.columns`

Understanding the dimensions and column names is fundamental. `df.shape` returns a tuple representing the dimensionality of the DataFrame (rows, columns), while `df.columns` provides a list of all column names.

*   **Number of rows**: Indicates the total number of records or observations in your dataset.
*   **Number of columns**: Represents the number of variables or features for each record.
*   **Column names**: Essential for referencing data within your DataFrame and understanding what each piece of information represents.

In [5]:
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")
print("Column names:")
print(df.columns.tolist())

Number of rows: 22
Number of columns: 9
Column names:
['EmployeeID', 'Name', 'Department', 'Age', 'Gender', 'Salary', 'Experience', 'City', 'PerformanceScore']


Knowing these basic pieces of information about a dataset is crucial before performing any cleaning function. It helps you:
*   Understand its overall size and structure.
*   Identify correct column names to avoid errors.

## 4. 📝 First Data Quality Inspection: Data Profiling

**Data Profiling** is the process of examining the data available in an information system (e.g. a database or a data warehouse) and collecting statistics and information about that data. It helps in understanding the data's structure, content, relationships, and quality. Think of it as taking an X-ray of your data!

We'll use several Pandas functions to perform our initial data profile.

### `df.info()`

The `.info()` method prints a concise summary of a DataFrame, including the index dtype and column dtypes, non-null values, and memory usage. This is an excellent first step for:

*   **Column names**: Verifying the names.
*   **Number of non-null values**: Quickly identifying columns with missing data (nulls).
*   **Data types (Dtypes)**: Understanding if columns are stored as integers, floats, objects (strings), etc., and checking for potential type mismatches.
*   **Potential missing values**: Any column with a 'non-null count' less than the total number of entries has missing values.

In [ ]:
print("DataFrame Information (df.info()):")
df.info()

DataFrame Information (df.info()):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   EmployeeID        21 non-null     float64
 1   Name              21 non-null     object 
 2   Department        21 non-null     object 
 3   Age               20 non-null     float64
 4   Gender            20 non-null     object 
 5   Salary            21 non-null     float64
 6   Experience        21 non-null     float64
 7   City              21 non-null     object 
 8   PerformanceScore  21 non-null     float64
dtypes: float64(5), object(4)
memory usage: 1.7+ KB


### `df.describe()`

The `.describe()` method generates descriptive statistics that summarize the central tendency, dispersion, and shape of a dataset's distribution, excluding `NaN` values. It's particularly useful for numerical columns and can reveal:

*   **Count**: Number of non-null observations.
*   **Mean**: The average value.
*   **Std (Standard Deviation)**: How spread out the values are.
*   **Min/Max**: The smallest and largest values, which can highlight outliers.
*   **25%/50%/75% (Quartiles)**: Give insights into the distribution of data.

This statistical summary can immediately alert you to potential issues like extreme outliers (e.g., an age of 590 years), vastly different scales, or values that don't make logical sense.

In [ ]:
print("Descriptive Statistics (df.describe()):")
display(df.describe())

Descriptive Statistics (df.describe()):


,EmployeeID,Age,Salary,Experience,PerformanceScore
count,21.000000,20.000000,2.100000e+01,21.000000,21.000000
mean,162.761905,70.600000,1.314052e+05,31.190476,123.142857
std,237.723769,122.829535,2.359396e+05,60.043833,187.364160
min,101.000000,25.000000,5.243600e+04,1.000000,71.000000
25%,107.000000,31.250000,6.273400e+04,11.000000,75.000000
50%,111.000000,45.500000,8.342800e+04,21.000000,85.000000
75%,116.000000,56.250000,1.031580e+05,28.000000,92.000000
max,1200.000000,590.000000,1.157200e+06,290.000000,940.000000


### `df.nunique()` and `df.dtypes`

*   `df.nunique()`: Returns the number of unique values for each column. This is very helpful for understanding categorical columns and spotting potential issues like inconsistent spellings or extra whitespace if the number of unique values is unexpectedly high.
*   `df.dtypes`: Returns a Series with the data type of each column. This helps confirm if columns are stored in the expected data type (e.g., numbers as numeric types, dates as datetime objects, text as objects).

In [ ]:
print("Number of Unique Values per Column (df.nunique()):")
display(df.nunique())

print("\nData Types per Column (df.dtypes):")
display(df.dtypes)

Number of Unique Values per Column (df.nunique()):


,0
EmployeeID,19
Name,19
Department,7
Age,15
Gender,4
Salary,19
Experience,15
City,7
PerformanceScore,14



Data Types per Column (df.dtypes):


,0
EmployeeID,float64
Name,object
Department,object
Age,float64
Gender,object
Salary,float64
Experience,float64
City,object
PerformanceScore,float64


## 5. ⚠️ Missing Values

Missing values are simply blank spaces or special markers (like `NaN` - Not a Number, or `None`) in your dataset where data should have been recorded but isn't. They are one of the most common data quality problems you'll encounter in real-world data.

### What are Missing Values?

Consider these examples:

```text
Customer Name = NaN
City = NaN
Sales = NaN
```

### Why do missing values cause problems?

Missing values can cause several issues during analysis:

*   **Skewed Statistics**: They can distort means, medians, and other statistical measures.
*   **Reduced Sample Size**: Many functions automatically drop rows with missing values, reducing your effective dataset size.
*   **Biased Results**: If the missingness is not random, it can introduce bias into your analysis.

### 🔎 Detect Missing Values

Pandas provides straightforward methods to identify missing values. It's important to know how to locate them and quantify their presence in your DataFrame.

#### Method 1: Using `df.isna().sum()`

The `.isna()` method returns a boolean DataFrame of the same shape as the original, indicating `True` where a value is missing (NaN) and `False` otherwise. Chaining `.sum()` to this result then counts the number of `True` values for each column, effectively giving us the total count of missing values per column.

In [ ]:
print("Total Missing Values per Column using `df.isna().sum()`:")
display(df.isna().sum())

Total Missing Values per Column using `df.isna().sum()`:


,0
EmployeeID,1
Name,1
Department,1
Age,2
Gender,2
Salary,1
Experience,1
City,1
PerformanceScore,1


## 6. 🧹 Missing Value Strategies

There is no one-size-fits-all solution for handling missing values. The best approach depends heavily on:

*   **Data type**: Is it numerical, categorical, or temporal?
*   **Business meaning**: What does a missing value in this specific column imply?
*   **Importance of the column**: Is this column critical for your analysis?
*   **Whether the missingness is random**: Is there a pattern to why data is missing?
*   **Whether the value can logically be inferred**: Can we reasonably guess what the missing value should be?

Let's explore some common strategies for dealing with missing data. For each strategy, we'll work with a *copy* of our DataFrame to demonstrate the effect without altering our original `df`.

### Approach 1: Remove Rows (`df.dropna()`)

The simplest way to deal with missing values is to remove the entire row that contains any missing data. This is done using the `dropna()` method.

`df.dropna()` will remove any row that has *at least one* missing value. You can control this behavior with arguments like `how='all'` (removes only if *all* values in the row are missing).

🔎 **Inspection**: Before dropping, let's see how many rows we would lose.

In [ ]:
df_before_drop = df.copy()
print(f"Original DataFrame shape: {df_before_drop.shape}")

# Count rows that would be dropped
rows_to_drop = df_before_drop.isnull().any(axis=1).sum()
print(f"Number of rows with at least one missing value: {rows_to_drop}")
print(f"Number of rows remaining after dropping: {df_before_drop.shape[0] - rows_to_drop}")

Original DataFrame shape: (22, 9)
Number of rows with at least one missing value: 7
Number of rows remaining after dropping: 15


🧹 **Cleaning**: Now, let's apply `dropna()` to a copy of our DataFrame.

In [6]:
df_dropped = df.copy() # Work on a copy
df_dropped.dropna(inplace=True) # Remove rows with any NaN value

print(f"DataFrame shape after dropping rows with any missing values: {df_dropped.shape}")

DataFrame shape after dropping rows with any missing values: (15, 9)


✅ **Validation**: Let's confirm that there are no more missing values in this new DataFrame.

In [ ]:
print("Missing values in df_dropped:")
display(df_dropped.isnull().sum())

Missing values in df_dropped:


,0
EmployeeID,0
Name,0
Department,0
Age,0
Gender,0
Salary,0
Experience,0
City,0
PerformanceScore,0


#### When is `dropna()` appropriate?

*   **Small percentage of missing data**: If only a tiny fraction of your rows have missing values, dropping them might not significantly impact your dataset size or statistical power.
*   **Random missingness**: If the missing values are randomly distributed and don't introduce bias.
*   **High data integrity requirement**: When precision is paramount, and imputation methods might introduce inaccuracies.
*   **No logical way to infer missing values**: If there's no reasonable way to guess what the missing data should be.

#### Dangers of `dropna()`:

*   **Loss of valuable data**: You might discard entire rows that contain valuable information in other columns, simply because one column is missing a value.
*   **Reduced sample size**: If many rows have missing values, your dataset can shrink dramatically, potentially affecting the reliability of your analysis.
*   **Bias**: If missingness is related to a specific characteristic (e.g., lower-income individuals tend not to fill out a certain field), dropping those rows can introduce bias into your remaining data.

### Approach 2: Fill with a Constant Value (`fillna()`)

Instead of removing data, we can replace missing values with a specific constant. This approach is suitable when you have a clear default or placeholder value for the missing data. The method used is `fillna()`.

#### Example 1: Filling numerical missing values with 0

Let's consider the `PerformanceScore` column. If a missing performance score logically implies a score of 0 (e.g., did not participate, or no record means no score), we can fill `NaN` with 0.

🔎 **Inspection**: Check current missing values in `PerformanceScore`.

In [22]:
df_fill_zero = df.copy() # Work on a copy
print(f"Missing values in PerformanceScore before filling: {df_fill_zero['PerformanceScore'].isnull().sum()}")

Missing values in PerformanceScore before filling: 1


🧹 **Cleaning**: Fill `NaN` values in `PerformanceScore` with 0.

In [23]:
df_fill_zero['PerformanceScore'].fillna(0, inplace=True)
print(f"Missing values in PerformanceScore after filling with 0: {df_fill_zero['PerformanceScore'].isnull().sum()}")

# Display relevant rows to show the change
print("\nRows where PerformanceScore was NaN (now 0):")
display(df_fill_zero[df['PerformanceScore'].isnull()][['Name', 'PerformanceScore']])

Missing values in PerformanceScore after filling with 0: 0

Rows where PerformanceScore was NaN (now 0):


/tmp/ipykernel_534/193500108.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_fill_zero['PerformanceScore'].fillna(0, inplace=True)


,Name,PerformanceScore
4,Eve Adams,0.0


#### Example 2: Filling categorical missing values with 'Unknown'

For categorical columns like `Gender` or `Department`, it often makes sense to fill missing values with 'Unknown' or 'Other' if the missingness itself provides information or if you don't want to make assumptions about the actual category.

🔎 **Inspection**: Check current missing values in `Gender`.

In [18]:
df_fill_unknown = df.copy() # Work on a copy
print(f"Missing values in Gender before filling: {df_fill_unknown['Gender'].isnull().sum()}")

Missing values in Gender before filling: 2


🧹 **Cleaning**: Fill `NaN` values in `Gender` with 'Unknown'.

In [19]:
df_fill_unknown['Gender'].fillna('Unknown', inplace=True)
print(f"Missing values in Gender after filling with 'Unknown': {df_fill_unknown['Gender'].isnull().sum()}")

# Display relevant rows to show the change
print("\nRows where Gender was NaN (now 'Unknown'):")
display(df_fill_unknown[df['Gender'].isnull()][['Name', 'Gender']])

Missing values in Gender after filling with 'Unknown': 0

Rows where Gender was NaN (now 'Unknown'):


/tmp/ipykernel_534/2895982362.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_fill_unknown['Gender'].fillna('Unknown', inplace=True)


,Name,Gender
8,Ivy Davis,Unknown
21,Ivy Davis,Unknown


#### When is filling with a constant appropriate?

*   **Meaningful placeholder**: When '0', 'Unknown', 'None', or a specific default value carries actual meaning for the missing data point.
*   **Categorical data**: Especially useful for categorical features where 'Unknown' can be treated as another category.
*   **Maintaining dataset size**: You don't lose any rows by using this method.

#### Dangers of filling with a constant:

*   **Distorting statistics**: Filling numerical missing values with 0 can significantly lower the mean and introduce a new mode at 0, altering the distribution of the data.
*   **Creating false information**: If 0 or 'Unknown' isn't truly representative, you're essentially inventing data.
*   **Misleading analysis**: Models might learn patterns from the imputed constant value, which might not reflect reality.

### Approach 3: Fill with Mean (`fillna(df['column'].mean())`)

For numerical columns, replacing missing values with the mean (average) of the existing values in that column is a common imputation strategy. This keeps the column's mean largely consistent.

Let's apply this to the `Salary` column, which has one missing value.

🔎 **Inspection**: Check current missing values and the mean of `Salary`.

In [ ]:
df_fill_mean = df.copy() # Work on a copy
print(f"Missing values in Salary before filling: {df_fill_mean['Salary'].isnull().sum()}")
salary_mean = df_fill_mean['Salary'].mean()
print(f"Mean Salary: {salary_mean:.2f}")

Missing values in Salary before filling: 1
Mean Salary: 131405.24


🧹 **Cleaning**: Fill `NaN` values in `Salary` with its mean.

In [ ]:
df_fill_mean['Salary'].fillna(salary_mean, inplace=True)
print(f"Missing values in Salary after filling with mean: {df_fill_mean['Salary'].isnull().sum()}")

# Display relevant rows to show the change
print("\nRows where Salary was NaN (now mean):")
display(df_fill_mean[df['Salary'].isnull()][['Name', 'Salary']])

Missing values in Salary after filling with mean: 0

Rows where Salary was NaN (now mean):


/tmp/ipykernel_3308/3726637003.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_fill_mean['Salary'].fillna(salary_mean, inplace=True)


,Name,Salary
17,Rachel Scott,131405.238095


#### When is mean imputation appropriate?

*   **Numerical data**: Only applicable to numerical features.
*   **Small amount of missing data**: If the proportion of missing values is small, the mean won't be drastically affected, and this is a simple, quick solution.
*   **Normally distributed data**: Mean is a good representation for normally distributed data.

#### Dangers of mean imputation:

*   **Reduces variance**: Replacing missing values with the mean makes the data points less varied, which can underestimate the true variability of the column.
*   **Distorts relationships**: Can weaken the correlation between the imputed column and other variables.
*   **Sensitive to outliers**: If the column contains outliers, the mean can be skewed, leading to an unrepresentative imputation value. The `Salary` column in our dataset actually has a high value (1157200.0) which might be an outlier and could skew the mean.

### Approach 4: Fill with Median (`fillna(df['column'].median())`)

Similar to mean imputation, median imputation replaces missing numerical values with the median (middle value) of the column. The median is less sensitive to outliers than the mean, making it a robust choice when your data is skewed or contains extreme values.

Let's apply this to the `Age` column, which has two missing values and also an apparent outlier (590.0) as seen in `df.describe()`.

🔎 **Inspection**: Check current missing values and the median of `Age`.

In [ ]:
df_fill_median = df.copy() # Work on a copy
print(f"Missing values in Age before filling: {df_fill_median['Age'].isnull().sum()}")
age_median = df_fill_median['Age'].median()
print(f"Median Age: {age_median:.0f}")

Missing values in Age before filling: 2
Median Age: 46


🧹 **Cleaning**: Fill `NaN` values in `Age` with its median.

In [ ]:
df_fill_median['Age'].fillna(age_median, inplace=True)
print(f"Missing values in Age after filling with median: {df_fill_median['Age'].isnull().sum()}")

# Display relevant rows to show the change
print("\nRows where Age was NaN (now median):")
display(df_fill_median[df['Age'].isnull()][['Name', 'Age']])

Missing values in Age after filling with median: 0

Rows where Age was NaN (now median):


/tmp/ipykernel_3308/2856882145.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_fill_median['Age'].fillna(age_median, inplace=True)


,Name,Age
8,Ivy Davis,45.5
21,Ivy Davis,45.5


#### When is median imputation appropriate?

*   **Numerical data**: Only applicable to numerical features.
*   **Skewed data or outliers**: Median is preferred over mean when the data distribution is skewed or contains outliers, as it provides a more representative central value.
*   **Small amount of missing data**: Similar to the mean, it works best when not too many values are missing.

#### Dangers of median imputation:

*   **Reduces variance**: Like mean imputation, it decreases the natural variability of the data.
*   **Distorts relationships**: Can weaken correlations between variables.
*   **May not be a 'real' value**: The median might not be an actual value present in the dataset, especially if there's an even number of observations.

### Approach 5: Fill with Mode (`fillna(df['column'].mode()[0])`)

Mode imputation replaces missing values with the most frequently occurring value in the column. This is particularly useful for categorical columns but can also be applied to numerical data where specific values recur often.

Let's apply this to the `Department` column, which has one missing value. We will also consider `EmployeeID` which is a float type but might be better filled by mode if it represents discrete IDs.

🔎 **Inspection**: Check current missing values and the mode of `Department` and `EmployeeID`.

In [ ]:
df_fill_mode = df.copy() # Work on a copy
print(f"Missing values in Department before filling: {df_fill_mode['Department'].isnull().sum()}")
department_mode = df_fill_mode['Department'].mode()[0]
print(f"Mode Department: {department_mode}")

print(f"\nMissing values in EmployeeID before filling: {df_fill_mode['EmployeeID'].isnull().sum()}")
employeeid_mode = df_fill_mode['EmployeeID'].mode()[0]
print(f"Mode EmployeeID: {employeeid_mode}")

Missing values in Department before filling: 1
Mode Department: HR

Missing values in EmployeeID before filling: 1
Mode EmployeeID: 109.0


🧹 **Cleaning**: Fill `NaN` values in `Department` and `EmployeeID` with their respective modes.

In [ ]:
df_fill_mode['Department'].fillna(department_mode, inplace=True)
df_fill_mode['EmployeeID'].fillna(employeeid_mode, inplace=True)

print(f"Missing values in Department after filling with mode: {df_fill_mode['Department'].isnull().sum()}")
print(f"Missing values in EmployeeID after filling with mode: {df_fill_mode['EmployeeID'].isnull().sum()}")

# Display relevant rows to show the change for Department
print("\nRows where Department was NaN (now mode):")
display(df_fill_mode[df['Department'].isnull()][['Name', 'Department']])

# Display relevant rows to show the change for EmployeeID
print("\nRows where EmployeeID was NaN (now mode):")
display(df_fill_mode[df['EmployeeID'].isnull()][['Name', 'EmployeeID']])

Missing values in Department after filling with mode: 0
Missing values in EmployeeID after filling with mode: 0

Rows where Department was NaN (now mode):


/tmp/ipykernel_3308/1699328735.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_fill_mode['Department'].fillna(department_mode, inplace=True)
/tmp/ipykernel_3308/1699328735.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inpl

,Name,Department
13,Noah Lewis,HR



Rows where EmployeeID was NaN (now mode):


,Name,EmployeeID
1,Bob Johnson,109.0


#### When is mode imputation appropriate?

*   **Categorical data**: This is its primary use case, as it's the only measure of central tendency for nominal data.
*   **Numerical data with specific recurring values**: If a numerical column has clear, frequently repeating values, mode imputation can be suitable.
*   **Maintaining dataset size**: No rows are removed.

#### Dangers of mode imputation:

*   **Multiple modes**: If a column has multiple modes (two or more values with the same highest frequency), `mode()[0]` will arbitrarily pick the first one, which might not be ideal.
*   **Over-representation**: Can artificially inflate the frequency of the mode value, potentially skewing analyses that rely on category counts or distributions.
*   **Introduces bias**: If the missingness is not random, imputing with the mode can reinforce existing biases.

## 7. ⚠️ Duplicate Records

Duplicate records are rows in your dataset that are identical or nearly identical to other rows, representing the same entity or observation more than once. They are a very common data quality problem, especially when data is collected from multiple sources or through systems that don't enforce uniqueness.

### What is a Duplicate?

A duplicate is a record that appears more than once when it should represent a unique observation. For example, if we have a list of employees, each `EmployeeID` should ideally be unique. If the same `EmployeeID` appears twice, it's a duplicate record.

### Why do duplicates cause problems?

Duplicates can severely impact your analysis:

*   **Inflate Totals**: If you sum `Salary` or count `EmployeeID`s, duplicates will lead to artificially higher numbers.
*   **Distort Averages**: Averages like mean `Age` or `PerformanceScore` will be skewed if the same person's data is included multiple times.
*   **Affect Counts**: Simple counts of customers, products, or transactions will be incorrect.
*   **Produce Incorrect Business Conclusions**: Any decision based on aggregated data will be flawed, leading to poor strategic choices.

In our `data_quality_summary` table, we saw that `EmployeeID` has 19 unique values for 22 rows, and `Name` also has 19 unique values. This strongly suggests that there are duplicate records in our dataset.

### 🔎 Detect Duplicates

Pandas provides excellent tools to identify duplicate rows. The primary method is `duplicated()`.

#### Using `df.duplicated()`

The `df.duplicated()` method returns a boolean Series indicating whether each row is a duplicate of a previous row. By default, it marks all duplicates as `True` except for the first occurrence.

Let's apply it to our DataFrame:

In [ ]:
print("Boolean Series indicating duplicate rows:")
display(df.duplicated().head())

Boolean Series indicating duplicate rows:


,0
0,False
1,False
2,False
3,False
4,False


To get a quick count of how many duplicate rows exist:

In [ ]:
num_duplicates = df.duplicated().sum()
print(f"Total number of duplicate rows (excluding first occurrence): {num_duplicates}")

Total number of duplicate rows (excluding first occurrence): 2


To display the actual duplicate records, we can filter the DataFrame using the boolean Series generated by `duplicated()`:

🔎 **Inspection**: Display the duplicate rows.

In [ ]:
print("Actual duplicate rows (excluding first occurrence):")
display(df[df.duplicated()])

Actual duplicate rows (excluding first occurrence):


,EmployeeID,Name,Department,Age,Gender,Salary,Experience,City,PerformanceScore
20,113.0,Mia Clark,HR,44.0,Female,57225.0,29.0,Paris,71.0
21,109.0,Ivy Davis,Marketing,NaN,NaN,106375.0,28.0,Tokyo,92.0


From the output, we can see two rows identified as duplicates:
*   Row 20: `EmployeeID` 113, `Mia Clark`, `HR`, `Age` 44, `Gender` Female, `Salary` 57225.0, `Experience` 29.0, `City` Paris, `PerformanceScore` 71.0
*   Row 21: `EmployeeID` 109, `Ivy Davis`, `Marketing`, `Age` NaN, `Gender` NaN, `Salary` 106375.0, `Experience` 28.0, `City` Tokyo, `PerformanceScore` 92.0

Let's look at the original DataFrame again to find their first occurrences:
*   Row 12: `EmployeeID` 113, `Mia Clark`, `HR`, `Age` 44.0, `Gender` Female, `Salary` 57225.0, `Experience` 29.0, `City` Paris, `PerformanceScore` 71.0 (This is identical to row 20)
*   Row 8: `EmployeeID` 109, `Ivy Davis`, `Marketing`, `Age` NaN, `Gender` NaN, `Salary` 106375.0, `Experience` 28.0, `City` Tokyo, `PerformanceScore` 92.0 (This is identical to row 21)

It seems we have two exact duplicate rows.

### Understanding `keep` parameter in `duplicated()` and `drop_duplicates()`

The `keep` parameter controls which duplicates (if any) are marked as `False` (original) or `True` (duplicate) in `duplicated()`, and which are kept in `drop_duplicates()`:

*   `keep='first'` (default): Marks all duplicates as `True` except for the first occurrence. When dropping, it keeps the first occurrence.
*   `keep='last'`: Marks all duplicates as `True` except for the last occurrence. When dropping, it keeps the last occurrence.
*   `keep=False`: Marks all duplicate occurrences as `True`. When dropping, it removes all occurrences of duplicates, keeping none.

## 8. 🧹 Remove Duplicates

Once identified, removing duplicate records is a critical cleaning step. The method `drop_duplicates()` is used for this purpose.

### Method 1: Remove exact duplicate rows

This method removes rows where *all* column values are identical to another row. This is the most common and safest way to remove obvious duplicates.

Let's create a copy of our original DataFrame and remove the exact duplicates.

🔎 **Inspection**: Check the number of duplicates before cleaning.

In [ ]:
df_no_exact_duplicates = df.copy()
print(f"Number of rows before removing exact duplicates: {df_no_exact_duplicates.shape[0]}")
print(f"Number of exact duplicate rows: {df_no_exact_duplicates.duplicated().sum()}")

Number of rows before removing exact duplicates: 22
Number of exact duplicate rows: 2


🧹 **Cleaning**: Remove exact duplicate rows.

In [ ]:
df_no_exact_duplicates.drop_duplicates(inplace=True)
print(f"Number of rows after removing exact duplicates: {df_no_exact_duplicates.shape[0]}")

Number of rows after removing exact duplicates: 20


✅ **Validation**: Verify that no exact duplicates remain.

In [ ]:
print(f"Number of exact duplicate rows remaining: {df_no_exact_duplicates.duplicated().sum()}")

Number of exact duplicate rows remaining: 0


### Method 2: Remove duplicates based on selected columns (`subset` parameter)

Sometimes, you might consider a record a duplicate if a specific combination of columns is identical, even if other columns differ. For example, if `EmployeeID` should uniquely identify an employee, then any rows with the same `EmployeeID` (even if other details are slightly different) might be considered duplicates that need resolution.

In our dataset, `EmployeeID` is a good candidate for a unique identifier. Let's assume that if `EmployeeID` is the same, it's the same person, and we should keep only one record. We'll use `keep='first'` to retain the first encountered record for each unique `EmployeeID`.

🔎 **Inspection**: Check for duplicates based on `EmployeeID`.

In [ ]:
df_subset_duplicates = df.copy()
print(f"Number of rows before removing duplicates by EmployeeID: {df_subset_duplicates.shape[0]}")
print(f"Number of duplicate EmployeeIDs (excluding first occurrence): {df_subset_duplicates.duplicated(subset=['EmployeeID']).sum()}")

# Display the rows that would be considered duplicates based on EmployeeID
print("\nRows considered duplicate based on 'EmployeeID':")
display(df_subset_duplicates[df_subset_duplicates.duplicated(subset=['EmployeeID'], keep=False)].sort_values('EmployeeID'))

Number of rows before removing duplicates by EmployeeID: 22
Number of duplicate EmployeeIDs (excluding first occurrence): 2

Rows considered duplicate based on 'EmployeeID':


,EmployeeID,Name,Department,Age,Gender,Salary,Experience,City,PerformanceScore
8,109.0,Ivy Davis,Marketing,NaN,NaN,106375.0,28.0,Tokyo,92.0
21,109.0,Ivy Davis,Marketing,NaN,NaN,106375.0,28.0,Tokyo,92.0
12,113.0,Mia Clark,HR,44.0,Female,57225.0,29.0,Paris,71.0
20,113.0,Mia Clark,HR,44.0,Female,57225.0,29.0,Paris,71.0


We can see that `EmployeeID` 109 and 113 appear twice. These are the same exact duplicate rows we identified earlier. If there were other columns that differed, `drop_duplicates(subset=['EmployeeID'])` would still remove one of them, based on the `keep` parameter.

🧹 **Cleaning**: Remove duplicates based on `EmployeeID`, keeping the first occurrence.

In [ ]:
df_unique_employeeid = df.copy()
df_unique_employeeid.drop_duplicates(subset=['EmployeeID'], keep='first', inplace=True)
print(f"Number of rows after removing duplicates by EmployeeID: {df_unique_employeeid.shape[0]}")

Number of rows after removing duplicates by EmployeeID: 20


✅ **Validation**: Verify that `EmployeeID` is now unique.

In [ ]:
print(f"Number of duplicate EmployeeIDs remaining: {df_unique_employeeid.duplicated(subset=['EmployeeID']).sum()}")
print(f"Number of unique EmployeeIDs: {df_unique_employeeid['EmployeeID'].nunique()}")

Number of duplicate EmployeeIDs remaining: 0
Number of unique EmployeeIDs: 19


#### Why are business rules important for duplicate removal?

Deciding what constitutes a 'unique' record often depends on the specific context and business logic. For instance:

*   A customer ID should almost always be unique.
*   An order ID should always be unique.
*   A product ID might be unique per store but not globally.

If you have a dataset of customer purchases, `CustomerID`, `OrderID`, and `ProductID` might all be important. A duplicate `CustomerID` might mean the same customer made two different purchases. A duplicate `OrderID` might indicate an error or a fraudulent transaction. A duplicate `ProductID` with the same `OrderID` usually means multiple items were bought in one transaction.

Always consult with subject matter experts to define the criteria for uniqueness before applying `drop_duplicates()` with a `subset`.

## 9. ⚠️ Whitespace Problems

Whitespace characters (spaces, tabs, newlines) at the beginning or end of text entries are a common, yet often overlooked, data quality issue. While they might appear invisible to the human eye, they are distinct characters that can cause problems in data analysis.

For example, consider these strings:

```text
"Amman"
" Amman"  <-- Leading space
"Amman "  <-- Trailing space
"  Amman  " <-- Both leading and trailing spaces
```

To Python and Pandas, these are four distinct values, even though they represent the same city. This can lead to:

*   **Inconsistent Categories**: 'New York' and 'New York ' will be treated as different cities, inflating unique counts and causing issues in grouping or filtering.
*   **Failed Merges/Joins**: If you try to merge datasets on a column with inconsistent whitespace, the merge might fail to find matches.
*   **Search/Filter Issues**: Searches for 'Amman' won't find ' Amman' or 'Amman '.

Let's inspect the `City` column, as we might expect city names to be consistent but our initial glance at `df.head()` and `df.tail()` (especially for 'New York') suggests some inconsistencies.

### 🔎 Detect Whitespace

We can detect leading or trailing whitespace programmatically. A simple way is to check if a string `startswith(' ')` or `endswith(' ')`.

#### Inspecting `City` column for whitespace

Let's check for leading and trailing spaces in the `City` column. We'll convert the column to string type first to ensure string methods work correctly and handle any potential `NaN` values gracefully.

In [ ]:
df_whitespace_check = df.copy()

# Convert City column to string type, filling NaN with empty strings for detection purposes
df_whitespace_check['City_str'] = df_whitespace_check['City'].astype(str).fillna('')

# Detect leading whitespace
leading_ws = df_whitespace_check['City_str'].str.startswith(' ')
print(f"Number of cities with leading whitespace: {leading_ws.sum()}")

# Detect trailing whitespace
trailing_ws = df_whitespace_check['City_str'].str.endswith(' ')
print(f"Number of cities with trailing whitespace: {trailing_ws.sum()}")

print("\nRows with leading whitespace:")
display(df_whitespace_check[leading_ws][['Name', 'City']])

print("\nRows with trailing whitespace:")
display(df_whitespace_check[trailing_ws][['Name', 'City']])

Number of cities with leading whitespace: 2
Number of cities with trailing whitespace: 2

Rows with leading whitespace:


,Name,City
9,Jack Miller,Berlin
15,NaN,NEW YORK



Rows with trailing whitespace:


,Name,City
9,Jack Miller,Berlin
15,NaN,NEW YORK


From the output, we can see 'Bob Johnson' and 'Karen Moore' have cities with leading spaces, and 'Amman' (though not in the data, it's an example), 'New York' and 'Berlin' from the dataset might have issues.

We can also use `str.len()` to compare string lengths before and after stripping whitespace to visually confirm the presence of extra spaces.

In [ ]:
df_len_check = df.copy()
# For non-null values in City, compare original length with stripped length
df_len_check['City_original_len'] = df_len_check['City'].astype(str).str.len()
df_len_check['City_stripped_len'] = df_len_check['City'].astype(str).str.strip().str.len()
df_len_check['Whitespace_Present'] = df_len_check['City_original_len'] != df_len_check['City_stripped_len']

print("Cities with length mismatch (indicating whitespace):")
display(df_len_check[df_len_check['Whitespace_Present']][['Name', 'City', 'City_original_len', 'City_stripped_len']])

Cities with length mismatch (indicating whitespace):


,Name,City,City_original_len,City_stripped_len
9,Jack Miller,Berlin,10,6
15,NaN,NEW YORK,12,8


This confirms that 'Bob Johnson' in ' City' and 'Karen Moore' in ' New York' have leading spaces, and 'Frank White' in 'New York ' and 'Olivia King' in 'Berlin ' have trailing spaces. 'Jack Miller' has 'Berlin  ', which seems to have both leading and trailing spaces.

Now that we've detected these issues, let's clean them up.

## 10. 🧹 Remove Leading and Trailing Whitespace

Pandas string methods are very powerful for cleaning text data. The `.str.strip()`, `.str.lstrip()`, and `.str.rstrip()` methods are specifically designed to handle whitespace.

### `str.strip()`

This is the most commonly used method. It removes *all* leading and trailing whitespace characters (spaces, tabs, newlines) from a string.

In [ ]:
df_stripped = df.copy() # Work on a copy

print("City column before stripping:")
display(df_stripped[df_stripped['City'].astype(str).str.contains(' ', regex=False)][['Name', 'City']])

# Apply strip to the City column
df_stripped['City'] = df_stripped['City'].astype(str).str.strip()

print("\nCity column after stripping:")
display(df_stripped[df_stripped['City'].astype(str).str.contains(' ', regex=False)][['Name', 'City']])

City column before stripping:


,Name,City
0,Alice Smith,New York
5,Frank White,New York
9,Jack Miller,Berlin
10,Karen Moore,New York
15,NaN,NEW YORK



City column after stripping:


,Name,City
0,Alice Smith,New York
5,Frank White,New York
10,Karen Moore,New York
15,NaN,NEW YORK


Notice how the extra spaces have been removed. We used `astype(str)` to ensure the method works even if there are `NaN` values, which are converted to the string 'nan' and then stripped (though this usually isn't an issue for 'nan'). For actual string `NaN` values, `str.strip()` will correctly return `NaN` if the value is originally `NaN`.

### `str.lstrip()`

Removes only leading whitespace characters from the left side of the string.

In [ ]:
df_lstripped = df.copy()
print("Original 'City' values with leading spaces:")
display(df_lstripped[df_lstripped['City'].astype(str).str.startswith(' ')][['Name', 'City']])

df_lstripped['City'] = df_lstripped['City'].astype(str).str.lstrip()
print("\n'City' values after lstrip():")
display(df_lstripped[df_lstripped['City'].astype(str).str.startswith(' ')][['Name', 'City']]) # Should be empty or only show values that still have leading spaces if they were originally multiple spaces long.

Original 'City' values with leading spaces:


,Name,City
9,Jack Miller,Berlin
15,NaN,NEW YORK



'City' values after lstrip():


,Name,City


### `str.rstrip()`

Removes only trailing whitespace characters from the right side of the string.

In [ ]:
df_rstripped = df.copy()
print("Original 'City' values with trailing spaces:")
display(df_rstripped[df_rstripped['City'].astype(str).str.endswith(' ')][['Name', 'City']])

df_rstripped['City'] = df_rstripped['City'].astype(str).str.rstrip()
print("\n'City' values after rstrip():")
display(df_rstripped[df_rstripped['City'].astype(str).str.endswith(' ')][['Name', 'City']]) # Should be empty or only show values that still have trailing spaces if they were originally multiple spaces long.

Original 'City' values with trailing spaces:


,Name,City
9,Jack Miller,Berlin
15,NaN,NEW YORK



'City' values after rstrip():


,Name,City


## 11. ⚠️ Inconsistent Text Formatting

Another common data quality problem with text data is inconsistent capitalization or formatting. Values that semantically represent the same category might appear differently due to varied casing or slight spelling variations.

For example, consider these variations for a city or department:

```text
Amman
amman
AMMAN
Marketing
marketing
MARKETING
```

To a computer, 'Marketing' and 'marketing' are distinct values. This can lead to:

*   **Inflated Unique Counts**: 'Marketing' and 'marketing' will be counted as two unique departments instead of one.
*   **Incorrect Grouping/Filtering**: When you try to group data by 'Department', you'll find separate groups for each casing variation, making analysis difficult.
*   **Difficulty in Merging**: Like whitespace, inconsistent casing can prevent successful merges between datasets.

Let's inspect our `Department` and `Gender` columns for such inconsistencies.

### 🔎 Detect Inconsistent Categories

The `value_counts()` method is an excellent tool for detecting inconsistent text formatting. It displays the frequency of each unique value in a Series, making it easy to spot variations that should be the same.

#### Inspecting `Department` column

🔎 **Inspection**: Let's check `value_counts()` for the `Department` column.

In [ ]:
print("Value counts for 'Department' before standardization:")
display(df['Department'].value_counts(dropna=False))

Value counts for 'Department' before standardization:


,count
Department,
IT,5
HR,5
Marketing,4
Sales,3
Finance,2
MARKETING,1
NaN,1
Sales,1


We can clearly see 'Marketing' (2 counts) and 'MARKETING' (1 count) which are the same department but with different casing. There's also one missing value (NaN). 'IT' and 'HR' also appear in different casings ('IT', ' IT'). We'll address the whitespace with stripping and then the casing.

#### Inspecting `Gender` column

🔎 **Inspection**: Let's check `value_counts()` for the `Gender` column.

In [ ]:
print("Value counts for 'Gender' before standardization:")
display(df['Gender'].value_counts(dropna=False))

Value counts for 'Gender' before standardization:


,count
Gender,
Female,9
Male,8
NaN,2
Female,2
MALE,1


Here, 'MALE' and 'Male' represent the same gender. We also have two missing values (NaN) and 'Female'.

### 🧹 Solutions for Inconsistent Text Formatting

The most common way to standardize text casing is to convert all entries to a uniform case (lowercase, uppercase, or title case). Before applying these, it's good practice to strip any leading/trailing whitespace, as demonstrated in the previous section.

#### First: Clean Whitespace (re-apply from previous section for demonstration)

It's crucial to remove whitespace *before* standardizing case, otherwise ` 'Marketing'` will become `' marketing'` instead of `'marketing'`.

In [ ]:
df_casing = df.copy()

# Clean whitespace from all object columns first
string_cols_to_clean = df_casing.select_dtypes(include='object').columns
for col in string_cols_to_clean:
    df_casing[col] = df_casing[col].astype(str).str.strip()

print("Value counts for 'Department' after whitespace stripping:")
display(df_casing['Department'].value_counts(dropna=False))
print("\nValue counts for 'Gender' after whitespace stripping:")
display(df_casing['Gender'].value_counts(dropna=False))

Value counts for 'Department' after whitespace stripping:


,count
Department,
IT,5
HR,5
Sales,4
Marketing,4
Finance,2
MARKETING,1
nan,1



Value counts for 'Gender' after whitespace stripping:


,count
Gender,
Female,11
Male,8
nan,2
MALE,1


Now that whitespace is handled, let's standardize the casing.

### Solution 1: Convert to Lowercase (`str.lower()`)

Converting all text to lowercase is a very common and effective way to standardize categories, especially when the case doesn't convey specific meaning.

In [ ]:
df_lower = df_casing.copy()

# Convert Department and Gender to lowercase
df_lower['Department'] = df_lower['Department'].str.lower()
df_lower['Gender'] = df_lower['Gender'].str.lower()

print("Value counts for 'Department' after converting to lowercase:")
display(df_lower['Department'].value_counts(dropna=False))
print("\nValue counts for 'Gender' after converting to lowercase:")
display(df_lower['Gender'].value_counts(dropna=False))

# ✅ Validation: Check if the original issues are resolved
print("\nUnique Department values after lowercase conversion:", df_lower['Department'].dropna().unique())
print("Unique Gender values after lowercase conversion:", df_lower['Gender'].dropna().unique())

Value counts for 'Department' after converting to lowercase:


,count
Department,
it,5
hr,5
marketing,5
sales,4
finance,2
nan,1



Value counts for 'Gender' after converting to lowercase:


,count
Gender,
female,11
male,9
nan,2



Unique Department values after lowercase conversion: ['it' 'hr' 'finance' 'marketing' 'sales' 'nan']
Unique Gender values after lowercase conversion: ['female' 'male' 'nan']


The categories 'Marketing' and 'MARKETING' are now uniformly 'marketing'. Similarly, 'MALE' and 'Male' are now 'male'. The counts have merged correctly!

### Solution 2: Convert to Uppercase (`str.upper()`)

Alternatively, you might choose to convert all text to uppercase. This is often done for codes or identifiers where uppercase is the standard, or simply as a stylistic choice for consistency.

In [ ]:
df_upper = df_casing.copy()

# Convert Department and Gender to uppercase
df_upper['Department'] = df_upper['Department'].str.upper()
df_upper['Gender'] = df_upper['Gender'].str.upper()

print("Value counts for 'Department' after converting to uppercase:")
display(df_upper['Department'].value_counts(dropna=False))
print("\nValue counts for 'Gender' after converting to uppercase:")
display(df_upper['Gender'].value_counts(dropna=False))

# ✅ Validation: Check if the original issues are resolved
print("\nUnique Department values after uppercase conversion:", df_upper['Department'].dropna().unique())
print("Unique Gender values after uppercase conversion:", df_upper['Gender'].dropna().unique())

Value counts for 'Department' after converting to uppercase:


,count
Department,
IT,5
HR,5
MARKETING,5
SALES,4
FINANCE,2
NAN,1



Value counts for 'Gender' after converting to uppercase:


,count
Gender,
FEMALE,11
MALE,9
NAN,2



Unique Department values after uppercase conversion: ['IT' 'HR' 'FINANCE' 'MARKETING' 'SALES' 'NAN']
Unique Gender values after uppercase conversion: ['FEMALE' 'MALE' 'NAN']


### Solution 3: Convert to Title Case (`str.title()`)

Title case capitalizes the first letter of each word in a string, and converts the rest to lowercase. This is often aesthetically pleasing for names, cities, or titles.

Let's apply this to the `City` column, as city names typically follow title case.

🔎 **Inspection**: First, let's look at `City` values after whitespace cleaning (from `df_casing`).

In [ ]:
print("Value counts for 'City' after whitespace stripping:")
display(df_casing['City'].value_counts(dropna=False))

Value counts for 'City' after whitespace stripping:


,count
City,
Paris,5
London,4
Tokyo,4
Berlin,4
New York,3
nan,1
NEW YORK,1


We see 'NEW YORK' and 'New York', which should be consistent. Let's apply title case.

🧹 **Cleaning**: Convert `City` to title case.

In [ ]:
df_title = df_casing.copy()

df_title['City'] = df_title['City'].str.title()

print("Value counts for 'City' after converting to title case:")
display(df_title['City'].value_counts(dropna=False))

# ✅ Validation: Check if the original issues are resolved
print("\nUnique City values after title case conversion:", df_title['City'].dropna().unique())

Value counts for 'City' after converting to title case:


,count
City,
Paris,5
New York,4
London,4
Tokyo,4
Berlin,4
Nan,1



Unique City values after title case conversion: ['New York' 'London' 'Paris' 'Tokyo' 'Berlin' 'Nan']


Now 'New York' and 'NEW YORK' are both standardized to 'New York'.

### 💡 Important: Standardization Should Follow Business Requirements

The choice of `lower()`, `upper()`, `title()`, or other string methods should not be based on personal preference alone. It must be driven by:

*   **Business Context**: What is the standard format used in your organization or for downstream systems?
*   **Readability**: Which format makes the data easiest to read and interpret for users?
*   **Consistency with Other Data Sources**: If you're combining data, match the casing of the primary or most reliable source.

Always clarify the desired standard with stakeholders or refer to data dictionaries.

## 12. ⚠️ Data Type Problems

Data types (e.g., integer, float, string, boolean, datetime) dictate how data is stored and how operations can be performed on them. Incorrect data types are a common data quality issue that can lead to errors in calculations, filters, and unexpected behavior in analysis.

### Inspecting Data Types: `df.dtypes`

We briefly looked at `df.dtypes` during our initial data inspection. Let's revisit it to identify potential issues.

🔎 **Inspection**:

In [ ]:
print("Current Data Types:")
display(df.dtypes)

Current Data Types:


,0
EmployeeID,float64
Name,object
Department,object
Age,float64
Gender,object
Salary,float64
Experience,float64
City,object
PerformanceScore,float64


From the output, we see:
*   `EmployeeID`, `Age`, `Salary`, `Experience`, `PerformanceScore` are all `float64`.
*   `Name`, `Department`, `Gender`, `City` are `object` (which typically means strings).

While `float64` is technically fine for numbers, `EmployeeID`, `Age`, `Experience`, and `PerformanceScore` are often represented as integers. Using `float64` for these can sometimes consume more memory or lead to minor precision issues if not handled carefully. More importantly, if a numerical column contains non-numeric entries, it might be incorrectly inferred as an `object` type by Pandas, preventing mathematical operations.

**Why incorrect data types are a problem:**

*   **Mathematical Operations**: You can't perform arithmetic on a column that is mistakenly an `object` type but contains numbers (e.g., summing salaries stored as strings).
*   **Filtering and Sorting**: Dates as strings won't sort chronologically. Numbers as strings won't sort numerically.
*   **Memory Usage**: `object` type columns (strings) often use more memory than numerical or datetime types.
*   **Analysis and Modeling**: Many statistical functions and machine learning algorithms require specific data types (e.g., numerical inputs).

**Examples of common data type problems:**

```text
"100" → string instead of number (e.g., Salary column)
"2026-01-15" → string instead of datetime (e.g., Date columns)
```

### 🧹 Numeric Conversion (`.astype()`)

You can convert columns to a specific numeric type using the `.astype()` method. This method directly casts the data in a column to the specified type.

#### Example: Converting `EmployeeID` and `Age` to Integer Type

Currently, `EmployeeID` and `Age` are `float64`. Since IDs and ages are typically whole numbers, converting them to an integer type (like `int64` or `Int64` for nullable integers) can be more appropriate. We should be careful with `NaN` values, as `int64` cannot directly store `NaN`. Pandas' `Int64` (nullable integer) type can handle this.

Let's use `df_no_exact_duplicates` (from our duplicate removal section) for this example to have a cleaner base, but we'll re-inspect the types and values.

In [ ]:
df_type_conversion = df_no_exact_duplicates.copy()

print("Original 'EmployeeID' and 'Age' dtypes:")
display(df_type_conversion[['EmployeeID', 'Age']].dtypes)

print("\nValues in 'EmployeeID' and 'Age' before conversion (including NaNs):")
display(df_type_conversion[['EmployeeID', 'Age']].sort_values('EmployeeID').head()) # Show some to see .0 and NaNs

# Convert EmployeeID and Age to nullable integer type (Int64)
df_type_conversion['EmployeeID'] = df_type_conversion['EmployeeID'].astype('Int64')
df_type_conversion['Age'] = df_type_conversion['Age'].astype('Int64')

print("\n'EmployeeID' and 'Age' dtypes after conversion:")
display(df_type_conversion[['EmployeeID', 'Age']].dtypes)

print("\nValues in 'EmployeeID' and 'Age' after conversion:")
display(df_type_conversion[['EmployeeID', 'Age']].sort_values('EmployeeID').head())

Original 'EmployeeID' and 'Age' dtypes:


,0
EmployeeID,float64
Age,float64



Values in 'EmployeeID' and 'Age' before conversion (including NaNs):


,EmployeeID,Age
0,101.0,26.0
2,103.0,57.0
3,104.0,47.0
4,105.0,42.0
5,106.0,29.0



'EmployeeID' and 'Age' dtypes after conversion:


,0
EmployeeID,Int64
Age,Int64



Values in 'EmployeeID' and 'Age' after conversion:


,EmployeeID,Age
0,101,26
2,103,57
3,104,47
4,105,42
5,106,29


The `Int64` dtype allows for integer values and `NaN` (which is now represented as `<NA>`).

#### Considerations for `.astype()`:

*   **Existing `NaN` values**: If a column has `NaN` values, you must use a nullable integer type like `Int64` (with a capital 'I') for integer conversions, as standard `int64` cannot represent `NaN`.
*   **Non-numeric values**: If the column contains non-numeric *strings* (e.g., 'abc') and you try to convert it directly to a numeric type, `.astype()` will raise an error. In such cases, you might need to handle those values first (e.g., by converting them to `NaN` manually) before using `.astype()`.

## 13. 🔄 Independent Cleaning Steps & Order of Operations

Throughout this notebook, we've demonstrated various cleaning operations independently. It's crucial to understand that data cleaning is often a sequence of distinct, independent steps. You typically don't combine all cleaning into one giant, chained operation or a single pipeline, especially when teaching or for clarity.

Here's a recommended sequence for independent cleaning steps:

1.  **Remove duplicates.**
2.  **Strip whitespace.**
3.  **Standardize text (e.g., casing).**
4.  **Fix data types.**
5.  **Handle missing values.**
6.  **Validate the result.**

For each step:
*   Explain the problem.
*   Detect the issue.
*   Make a decision.
*   Apply the cleaning operation.
*   Validate the result.
*   Display the relevant before-and-after values.

### Why the Order of Operations Matters 💡

The order in which you apply cleaning operations can significantly impact the outcome. Some operations are prerequisites for others, or their effects can influence subsequent steps.

Consider this typical sequence and why it's often preferred:

1.  **Whitespace Cleaning first**: It's often best to remove leading/trailing whitespace (`.str.strip()`) *before* standardizing text casing or checking for duplicates. Why? Because `" Marketing"` and `"Marketing"` are different strings due to the space. If you convert to lowercase first, you get `" marketing"` and `"marketing"`, which are still different. Cleaning whitespace resolves this initial inconsistency.

2.  **Standardizing Text Casing**: After whitespace, standardizing text (`.str.lower()`, `.str.upper()`, `.str.title()`) ensures that semantically identical categories (e.g., 'Marketing', 'MARKETING', 'marketing') are treated as one. This directly impacts duplicate detection and unique value counts.

3.  **Duplicate Detection and Removal**: If you remove duplicates *before* cleaning whitespace and standardizing casing, you might miss duplicates. For example, if you have `"Marketing "` and `"MARKETING"`, they won't be identified as duplicates until after they are both cleaned to `"marketing"`. Therefore, cleaning text *before* checking for duplicates is critical for accurate identification.

4.  **Data Type Correction**: Often, fixing data types requires clean, consistent data. For instance, converting a column to numeric type works best when there are no stray non-numeric characters (which might be caused by inconsistent text or whitespace).

5.  **Handling Missing Values**: Imputation strategies for missing values might depend on the distribution or categories of the cleaned data. For example, calculating a mean or mode for imputation makes more sense once other data quality issues are addressed.

This structured approach ensures that each cleaning step builds on the previous one, leading to a more accurate and robust clean dataset. However, remember that the specific order can still vary based on the unique characteristics and problems of your dataset.